File: desktop/desktop.go
// Package desktop provides an optional, host-independent desktop frontend bridge.
// Native Wails packages belong in the application's host module, never here.
package desktop

import (
	"bytes"
	"context"
	_ "embed"
	"encoding/json"
	"errors"
	"io"
	"math"
	"strings"
	"sync"
	"time"

	"github.com/monstercameron/GoWebComponents/v6/interop"
)

// ProtocolVersion identifies the synchronous transport envelope contract.
const ProtocolVersion = 1

// RequestTimeout is the default call ceiling, including context.Background calls.
const RequestTimeout = 30 * time.Second

// MaximumRequestTimeout bounds explicitly extended interactive calls.
const MaximumRequestTimeout = 10 * time.Minute

// BootstrapSource is the external ES module to serve as desktop.js before Wasm.
//
//go:embed desktop.js
var BootstrapSource string

// Capabilities describes an installed bridge, not an authorization credential.
type Capabilities struct {
	Protocol    int      `json:"protocol"`
	Platform    string   `json:"platform"`
	HostVersion string   `json:"hostVersion"`
	Methods     []string `json:"methods"`
	Topics      []string `json:"topics"`
	// Features advertises non-RPC capabilities such as installed native menus.
	// It is feature discovery, never authorization; RPC features still require methods.
	Features []Feature `json:"features,omitempty"`
}

// Reply is a synchronously polled request/event envelope; Data is JSON, not JS handles.
type Reply struct {
	Done    bool              `json:"done"`
	Data    json.RawMessage   `json:"data"`
	Code    interop.ErrorCode `json:"code,omitempty"`
	Message string            `json:"message,omitempty"`
}

// Transport can be injected in native tests. Methods must return promptly and
// serialize access internally. Cancel must forget requests even if native work stalls.
type Transport interface {
	Capabilities() (Capabilities, error)
	Start(string, json.RawMessage) (string, error)
	Poll(string) (Reply, error)
	Cancel(string) error
	Listen(string) (string, error)
	Next(string) (Reply, error)
	Unlisten(string) error
}

// Client owns calls and subscriptions; its zero value is predictably unavailable.
type Client struct{ parseTransport Transport }

// NewClient constructs a client with an explicit transport for testing or embedding.
func NewClient(parseTransport Transport) Client { return Client{parseTransport: parseTransport} }

// GetCapabilities validates the installed protocol before callers enable controls.
func (parseClient Client) GetCapabilities() (Capabilities, error) {
	if parseClient.parseTransport == nil {
		return Capabilities{}, getError("capabilities", interop.CodeUnavailable, errors.New("desktop host unavailable"))
	}
	parseCaps, parseErr := parseClient.parseTransport.Capabilities()
	if parseErr == nil && parseCaps.Protocol != ProtocolVersion {
		parseErr = getError("capabilities", interop.CodeInvalid, errors.New("desktop protocol mismatch"))
	}
	return parseCaps, parseErr
}

// Call decodes one registered method's JSON response. Use strings for wide integer
// identifiers, base64 strings for bytes and RFC3339 strings for time values.
// Cancellation is cooperative and cannot undo completed native writes.
func Call[T any](parseContext context.Context, parseClient Client, parseMethod string, parseArgs ...any) (T, error) {
	return CallWithTimeout[T](parseContext, parseClient, RequestTimeout, parseMethod, parseArgs...)
}

// CallWithTimeout opts one call into a positive ceiling of at most ten minutes.
// Earlier context deadlines still win. Use longer waits only for interactive native
// operations; cancellation cannot dismiss every OS dialog or undo completed writes.
func CallWithTimeout[T any](parseContext context.Context, parseClient Client, parseTimeout time.Duration, parseMethod string, parseArgs ...any) (T, error) {
	var parseZero T
	if parseTimeout <= 0 || parseTimeout > MaximumRequestTimeout {
		return parseZero, getError(parseMethod, interop.CodeInvalid, errors.New("call timeout must be positive and at most ten minutes"))
	}
	parseContext, parseCancel := getCallContext(parseContext, parseTimeout)
	defer parseCancel()
	if parseErr := parseContext.Err(); parseErr != nil {
		return parseZero, getContextError(parseMethod, parseErr)
	}
	parseCaps, parseErr := parseClient.GetCapabilities()
	if parseErr != nil {
		return parseZero, parseErr
	}
	if !hasName(parseCaps.Methods, parseMethod) {
		return parseZero, getError(parseMethod, interop.CodeMissingExport, errors.New("method not registered"))
	}
	if parseArgs == nil {
		parseArgs = []any{}
	}
	parseData, parseErr := json.Marshal(parseArgs)
	if parseErr != nil {
		return parseZero, getError(parseMethod, interop.CodeEncode, parseErr)
	}
	if parseErr = validateWireNumbers(parseData); parseErr != nil {
		return parseZero, getError(parseMethod, interop.CodeEncode, parseErr)
	}
	if parseErr = parseContext.Err(); parseErr != nil {
		return parseZero, getContextError(parseMethod, parseErr)
	}
	parseID, parseErr := parseClient.parseTransport.Start(parseMethod, parseData)
	if parseErr != nil {
		return parseZero, parseErr
	}
	// Cleanup forgets the request even when forwarding native cancellation fails.
	defer func() { _ = parseClient.parseTransport.Cancel(parseID) }()
	parseTicker := time.NewTicker(5 * time.Millisecond)
	defer parseTicker.Stop()
	for {
		if parseErr = parseContext.Err(); parseErr != nil {
			return parseZero, getContextError(parseMethod, parseErr)
		}
		parseReply, parsePollErr := parseClient.parseTransport.Poll(parseID)
		// A transport can synchronously trigger cancellation while producing its final reply.
		if parseErr = parseContext.Err(); parseErr != nil {
			return parseZero, getContextError(parseMethod, parseErr)
		}
		if parsePollErr != nil {
			return parseZero, parsePollErr
		}
		if parseReply.Done {
			if parseReply.Code != "" {
				return parseZero, getError(parseMethod, parseReply.Code, errors.New(parseReply.Message))
			}
			if parseErr = json.Unmarshal(parseReply.Data, &parseZero); parseErr != nil {
				return parseZero, getError(parseMethod, interop.CodeDecode, parseErr)
			}
			return parseZero, nil
		}
		select {
		case <-parseContext.Done():
			return parseZero, getContextError(parseMethod, parseContext.Err())
		case <-parseTicker.C:
		}
	}
}

// getCallContext applies a call ceiling without extending an earlier caller deadline.
func getCallContext(parseContext context.Context, parseTimeout time.Duration) (context.Context, context.CancelFunc) {
	if parseContext == nil {
		parseContext = context.Background()
	}
	return context.WithTimeout(parseContext, parseTimeout)
}

// validateWireNumbers rejects numeric magnitudes that lose integer precision in JavaScript.
func validateWireNumbers(parseData []byte) error {
	parseDecoder := json.NewDecoder(bytes.NewReader(parseData))
	parseDecoder.UseNumber()
	for {
		parseToken, parseErr := parseDecoder.Token()
		if parseErr == io.EOF {
			return nil
		}
		if parseErr != nil {
			return parseErr
		}
		if parseNumber, isNumber := parseToken.(json.Number); isNumber {
			parseValue, parseErr := parseNumber.Float64()
			if parseErr != nil || math.Abs(parseValue) > 9007199254740991 {
				return errors.New("numeric payload exceeds JavaScript safe range; encode wide numbers as strings")
			}
		}
	}
}

// Subscribe delivers latest-value events at most once per polling tick (16 ms).
// Bursts coalesce per subscription, so this API is for state/progress, not audit logs.
// Cancel is idempotent; a handler already executing may finish after cancellation.
func Subscribe[T any](parseContext context.Context, parseClient Client, parseTopic string, parseHandler func(T, error)) (func(), error) {
	if parseContext == nil {
		parseContext = context.Background()
	}
	if parseErr := parseContext.Err(); parseErr != nil {
		return nil, getContextError(parseTopic, parseErr)
	}
	if parseHandler == nil || !strings.Contains(parseTopic, ".") {
		return nil, getError(parseTopic, interop.CodeInvalid, errors.New("namespaced topic and handler required"))
	}
	parseCaps, parseErr := parseClient.GetCapabilities()
	if parseErr != nil {
		return nil, parseErr
	}
	if !hasName(parseCaps.Topics, parseTopic) {
		return nil, getError(parseTopic, interop.CodeMissingExport, errors.New("topic not registered"))
	}
	parseID, parseErr := parseClient.parseTransport.Listen(parseTopic)
	if parseErr != nil {
		return nil, parseErr
	}
	parseContext, parseCancel := context.WithCancel(parseContext)
	var parseOnce sync.Once
	parseStop := func() { parseOnce.Do(func() { parseCancel(); _ = parseClient.parseTransport.Unlisten(parseID) }) }
	go func() {
		defer interop.RecoverContainedPanic("desktop subscription")
		defer parseStop()
		parseTicker := time.NewTicker(16 * time.Millisecond)
		defer parseTicker.Stop()
		for {
			select {
			case <-parseContext.Done():
				return
			case <-parseTicker.C:
			}
			parseReply, parsePollErr := parseClient.parseTransport.Next(parseID)
			if parseContext.Err() != nil {
				return
			}
			if parsePollErr != nil {
				var parseZero T
				parseHandler(parseZero, parsePollErr)
				return
			}
			if !parseReply.Done {
				continue
			}
			var parseValue T
			if parseReply.Code != "" {
				parsePollErr = getError(parseTopic, parseReply.Code, errors.New(parseReply.Message))
			} else {
				parsePollErr = json.Unmarshal(parseReply.Data, &parseValue)
				if parsePollErr != nil {
					parsePollErr = getError(parseTopic, interop.CodeDecode, parsePollErr)
				}
			}
			parseHandler(parseValue, parsePollErr)
		}
	}()
	return parseStop, nil
}

// hasName checks exact allowlisted names without wildcard expansion.
func hasName(parseNames []string, parseName string) bool {
	for _, parseItem := range parseNames {
		if parseItem == parseName {
			return true
		}
	}
	return false
}

// getError preserves the shared interop error classification.
func getError(parseTarget string, parseCode interop.ErrorCode, parseErr error) error {
	return &interop.Error{Op: "desktop", Target: parseTarget, Code: parseCode, Err: parseErr}
}

// getContextError distinguishes deadlines from explicit cancellation.
func getContextError(parseTarget string, parseErr error) error {
	parseCode := interop.CodeCancelled
	if errors.Is(parseErr, context.DeadlineExceeded) {
		parseCode = interop.CodeTimeout
	}
	return getError(parseTarget, parseCode, parseErr)
}

File: desktop/desktop.js
// createDesktopTransport adapts explicitly registered generated Wails exports.
// No Go js.Func is captured by a promise; registries and event queues are bounded.
export function createDesktopTransport({ methods, topics = [], features = [], events, platform, hostVersion, limit = 256 }) {
  if (!Number.isInteger(limit) || limit < 1 || limit > 4096) throw new Error("Invalid desktop registry limit");
  const getMethods = new Map(Object.entries(methods));
  const getTopics = new Set(topics);
  if (!Array.isArray(features) || features.length > 64 || features.some(parseFeature => typeof parseFeature !== "string" || !parseFeature || parseFeature.length > 128)) throw new Error("Invalid desktop feature capabilities");
  const getFeatures = new Set(features);
  for (const [parseName, parseMethod] of getMethods) {
    if (!parseName || typeof parseMethod !== "function") throw new Error("Desktop methods require names and callable exports");
  }
  for (const parseTopic of getTopics) {
    if (typeof parseTopic !== "string" || !parseTopic.includes(".")) throw new Error("Desktop topics must be namespaced");
  }
  if (getTopics.size && typeof events?.On !== "function") throw new Error("Desktop event runtime unavailable");
  const getRequests = new Map();
  const getListeners = new Map();
  let getSequence = 0;
  let isClosed = false;
  const getOK = (parseData = null) => JSON.stringify({ data: parseData });
  // Host errors may contain throwing getters; error reporting must itself remain total.
  const getMessage = (parseError) => {
    try { return String(parseError?.message || parseError); }
    catch { return "Desktop host error could not be serialized"; }
  };
  const getError = (parseCode, parseMessage) => JSON.stringify({ code: parseCode, message: getMessage(parseMessage) });
  const getReply = (parseValue) => {
    try { return { done: true, data: JSON.parse(JSON.stringify(parseValue ?? null, (parseKey, parseItem) => {
      if (typeof parseItem === "number" && (!Number.isFinite(parseItem) || (Number.isInteger(parseItem) && !Number.isSafeInteger(parseItem)))) throw new Error("Numeric payload exceeds JavaScript safe range; use strings");
      return parseItem;
    })) }; }
    catch (parseError) { return { done: true, code: "decode", message: getMessage(parseError) }; }
  };
  const cancelPromise = (parsePromise) => {
    try { if (typeof parsePromise?.cancel === "function") Promise.resolve(parsePromise.cancel()).catch(() => {}); }
    catch { /* The owned registry is already released, even if host cancellation throws. */ }
  };
  // Cancellation errors are consumed, but cancellation does not roll back writes.
  const cancelRequest = (parseID) => {
    const parseEntry = getRequests.get(parseID);
    getRequests.delete(parseID);
    if (parseEntry && !parseEntry.reply.done) cancelPromise(parseEntry.promise);
  };
  const stopListener = (parseID) => {
    const parseEntry = getListeners.get(parseID);
    getListeners.delete(parseID);
    if (parseEntry) parseEntry.stop();
  };
  const getTransport = {
    capabilities() {
      return isClosed ? getError("disposed", "Desktop window closed") : getOK({ protocol: 1, platform, hostVersion, methods: [...getMethods.keys()], topics: [...getTopics], features: [...getFeatures] });
    },
    start(parseMethod, parseJSON) {
      if (isClosed) return getError("disposed", "Desktop window closed");
      if (!getMethods.has(parseMethod)) return getError("missing_export", "Method not registered");
      if (getRequests.size >= limit) return getError("quota_exceeded", "Desktop request limit reached");
      let parseArgs;
      try { parseArgs = JSON.parse(parseJSON); if (!Array.isArray(parseArgs)) throw new Error("Arguments must be an array"); }
      catch (parseError) { return getError("encode", parseError); }
      const parseID = String(++getSequence);
      try {
        const parsePromise = getMethods.get(parseMethod)(...parseArgs);
        if (isClosed) {
          Promise.resolve(parsePromise).catch(() => {});
          cancelPromise(parsePromise);
          return getError("disposed", "Desktop window closed during method invocation");
        }
        getRequests.set(parseID, { promise: parsePromise, reply: { done: false } });
        Promise.resolve(parsePromise).then(
          (parseValue) => { const parseEntry = getRequests.get(parseID); if (parseEntry) { parseEntry.reply = getReply(parseValue); parseEntry.promise = null; } },
          (parseError) => { const parseEntry = getRequests.get(parseID); if (parseEntry) { parseEntry.reply = { done: true, code: "remote_error", message: getMessage(parseError) }; parseEntry.promise = null; } }
        );
        return getOK(parseID);
      } catch (parseError) { cancelRequest(parseID); return getError("remote_error", parseError); }
    },
    poll(parseID) {
      if (isClosed) return getError("disposed", "Desktop window closed");
      const parseEntry = getRequests.get(parseID);
      return parseEntry ? getOK(parseEntry.reply) : getError("disposed", "Request released");
    },
    cancel(parseID) { cancelRequest(parseID); return getOK(); },
    listen(parseTopic) {
      if (isClosed) return getError("disposed", "Desktop window closed");
      if (!getTopics.has(parseTopic)) return getError("missing_export", "Topic not registered");
      if (getListeners.size >= limit) return getError("quota_exceeded", "Desktop subscription limit reached");
      const parseID = String(++getSequence);
      const parseEntry = { reply: { done: false }, stop: () => {} };
      getListeners.set(parseID, parseEntry);
      try {
        parseEntry.stop = events.On(parseTopic, (parseEvent) => {
          if (!getListeners.has(parseID)) return;
          try { parseEntry.reply = getReply(parseEvent.data); }
          catch (parseError) { parseEntry.reply = { done: true, code: "decode", message: getMessage(parseError) }; }
        });
        if (typeof parseEntry.stop !== "function") throw new Error("Desktop event runtime did not return unsubscribe");
        if (isClosed) {
          parseEntry.stop();
          return getError("disposed", "Desktop window closed during event registration");
        }
      }
      catch (parseError) { getListeners.delete(parseID); return getError("remote_error", parseError); }
      return getOK(parseID);
    },
    next(parseID) {
      if (isClosed) return getError("disposed", "Desktop window closed");
      const parseEntry = getListeners.get(parseID);
      if (!parseEntry) return getError("disposed", "Subscription released");
      const parseReply = parseEntry.reply;
      parseEntry.reply = { done: false };
      return getOK(parseReply);
    },
    unlisten(parseID) { try { stopListener(parseID); return getOK(); } catch (parseError) { return getError("remote_error", parseError); } },
    close() {
      if (isClosed) return;
      isClosed = true;
      for (const parseID of getRequests.keys()) cancelRequest(parseID);
      for (const parseID of getListeners.keys()) { try { stopListener(parseID); } catch (parseError) { console.error("Desktop unsubscribe failed", parseError); } }
    },
    // stats exposes counts for lifecycle tests without exposing request payloads.
    stats() { return { requests: getRequests.size, subscriptions: getListeners.size }; }
  };
  return getTransport;
}
